flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.scala

```diff
@@ -101,7 +101,7 @@ class PushProjectIntoTableSourceScanRule extends RelOptRule(
       .getTypeFactory
       .asInstanceOf[FlinkTypeFactory]
     val newSourceRowType = TableSourceUtil.getSourceRowType(flinkTypeFactory,
-      oldTableSource.getTableSchema,
+      tableSourceTable.catalogTable.getSchema,
```

Comment:
I really doubt this. The source row type should not use `catalogTable#tableSchema` which may contain computed column. That's why I was suggesting to put the row type logic in `TableSourceTable`. This is very easy to make mistake.